File: lib/src/taglib_ffi.dart
import 'dart:async';
import 'dart:ffi';
import 'dart:io';
import 'dart:isolate';

import 'package:ffi/ffi.dart';
import 'package:flutter/services.dart';
import 'package:isolate_pool_2/isolate_pool_2.dart';

import 'taglib_ffi_bindings_generated.dart';

class Metadata {
  Metadata({
    required this.title,
    required this.artist,
    required this.album,
    required this.track,
    required this.year,
    required this.genre,
    required this.comment,
    required this.sampleRate,
    required this.bitrate,
    required this.channels,
    required this.length,
    this.albumArtist,
    this.albumTotalTrack,
    this.lyrics,
    this.albumCover,
  });

  final String? title;
  final String? artist;
  final String? album;
  final String? albumArtist;
  final int? track;
  final int? albumTotalTrack;
  final int? year;
  final String? genre;
  final String? comment;
  final int? sampleRate;
  final int? bitrate;
  final int? channels;
  final int? length;
  final String? lyrics;
  final Uint8List? albumCover;
}

late final _TagLib _taglib;

var _initialized = false;

Future<void> initialize({int? isolateCount}) async {
  if (_initialized) {
    return;
  }
  _taglib = _TagLib(isolateCount);
  await _taglib.init();
  _initialized = true;
}

void checkInitialization() {
  if (!_initialized) {
    throw Exception(
        'TagLibFfi: Need to call "TagLib.initialize()" before calling TagLibFfi APIs');
  }
}

class _ReadMetadataJob extends PooledJob<Metadata?> {
  _ReadMetadataJob(this.filePath);

  final String filePath;

  @override
  Future<Metadata?> job() async {
    if (filePath.isEmpty) {
      return null;
    }
    final p = ReceivePort();
    Isolate.spawn(_readMetadata, <dynamic>[p.sendPort, filePath]);
    return await p.first;
  }
}

class _TagLib {
  _TagLib([int? isolateCount]) {
    this.isolateCount = isolateCount ?? 4;
    pool = IsolatePool(this.isolateCount);
  }

  Future<void> init() async {
    await pool.start();
  }

  late final int isolateCount;
  late final IsolatePool pool;

  Future<Metadata?> readMetadata(String filePath) async {
    return pool.scheduleJob(_ReadMetadataJob(filePath));
  }
}

Future<Metadata?> readMetadata(String filePath) async {
  return _taglib.readMetadata(filePath);
}

Future<dynamic> _readMetadata(List<dynamic> params) async {
  final p = params[0] as SendPort;
  final filePath = params[1] as String;
  Pointer<ID3v2Tag> originalTag = nullptr;
  late final TaglibFfiBindings meipuru;
  try {
    meipuru = TaglibFfiBindings(
      Platform.isWindows
          ? DynamicLibrary.open('taglib_ffi.dll')
          : DynamicLibrary.open('libtaglib_ffi.so'),
    );
    if (Platform.isWindows) {
      // Use wchar_t as parameter on Windows.
      // FIXME: Fix multiplatform compatibility.
      // final Pointer<WChar> tagFileName = filePath.toNativeUtf16().cast();
      // originalTag = meipuru.MeipuruReadID3v2TagW(tagFileName);
      // try {
      //   final locale = await Get.find<LocaleService>().getLocale();
      //   print('AAAA LOCALE: ${locale}');
      //   if (locale.languageCode == 'zh' && locale.countryCode == 'CN') {
      //     //if (locale != null && locale == 'zh_CN') {
      //     // tagFileName = filePath.toNativeGbk().cast();
      //     tagFileName = filePath.toNativeUtf8().cast();
      //   } else {
      //     tagFileName = filePath.toNativeUtf8().cast();
      //   }
      // } catch (e) {
      //   print('Error in _readMetadataEx getting tagFileName: $e');
      //   tagFileName = filePath.toNativeUtf8().cast();
      // }
    } else {
      final Pointer<Char> tagFileName = filePath.toNativeUtf8().cast();
      originalTag = meipuru.readID3v2Tag(tagFileName);
    }
    if (originalTag.address == nullptr.address) {
      print('FFI returned nullptr in meipuru.MeipuruReadID3v2Tag');
      return Isolate.exit(p);
    }

    final id3v2Tag = originalTag.cast<ID3v2Tag>().ref;
    final metaData = Metadata(
      title: id3v2Tag.title.cast<Utf8>().toDartString(),
      artist: id3v2Tag.artist.cast<Utf8>().toDartString(),
      album: id3v2Tag.albumTitle.cast<Utf8>().toDartString(),
      track: id3v2Tag.track,
      year: id3v2Tag.year,
      genre: id3v2Tag.genre.cast<Utf8>().toDartString(),
      comment: id3v2Tag.comment.cast<Utf8>().toDartString(),
      sampleRate: id3v2Tag.sampleRate,
      bitrate: id3v2Tag.bitRate,
      channels: id3v2Tag.channels,
      length: id3v2Tag.length,
      albumArtist: id3v2Tag.albumArtist.cast<Utf8>().toDartString(),
      albumTotalTrack: id3v2Tag.albumTotalTrack,
      lyrics: null,
      albumCover: id3v2Tag.albumCoverLength > 0
          ? id3v2Tag.albumCover
              .cast<Uint8>()
              .asTypedList(id3v2Tag.albumCoverLength)
          : null,
    );
    /*
      id3v2Tag.lyricsLength > 0
            ? id3v2Tag.lyrics
                .cast<Utf8>()
                .toDartString(length: id3v2Tag.lyricsLength)
            : null

       id3v2Tag.albumCoverLength > 0
            ? id3v2Tag.albumCover
                .cast<Uint8>()
                .asTypedList(id3v2Tag.albumCoverLength)
            : null
       */
    meipuru.freeID3v2Tag(originalTag);
    return Isolate.exit(p, metaData);
  } catch (e) {
    if (originalTag != nullptr) {
      meipuru.freeID3v2Tag(originalTag);
    }
    print('Error in readMetadataEx: $e');
    return Isolate.exit(p, null);
  }
}

File: packages/taglib_ffi_dart/src/taglib_ffi.cpp
#include "taglib_ffi.h"

#include "src/Reader.hpp"

struct Id3v2TagOwner {
    // On heap.
    const FFI::ID3v2::ID3v2 *tag;
};

const ID3v2Tag *readID3v2Tag(const char *filePath, bool readImage) {
    const auto readerOption = FFI::ReaderOption{
        FFI::Util::Encoding::Utf8,
        FFI::Util::LogLevel::None,
        true,
    };
    const auto reader = FFI::Reader(readerOption);
    const auto id3v2Tag = reader.readID3v2TagFromFile(filePath);

    if (id3v2Tag == nullptr) {
        return nullptr;
    }

    const auto tag = new ID3v2Tag{
        id3v2Tag->filePath.c_str(),
        id3v2Tag->fileName.c_str(),
        id3v2Tag->title.c_str(),
        nullptr, // Aritst
        id3v2Tag->artist.size(),
        id3v2Tag->albumTitle.c_str(),
        nullptr, // Album artist
        id3v2Tag->albumArtist.size(),
        id3v2Tag->year,
        id3v2Tag->track,
        id3v2Tag->albumTotalTrack,
        id3v2Tag->genre.c_str(),
        id3v2Tag->comment.c_str(),
        id3v2Tag->bitRate,
        id3v2Tag->sampleRate,
        id3v2Tag->channels,
        id3v2Tag->lengthInSeconds,
        id3v2Tag->lengthInMilliseconds,
        nullptr, // Picture
        id3v2Tag->pictures.size(),
        id3v2Tag->lyrics.c_str(),
        new Id3v2TagOwner{id3v2Tag},
    };
    // FIXME:
    tag->artist = (const char **)malloc(sizeof(char *) * id3v2Tag->artist.size());
    for (auto i = 0; i < id3v2Tag->artist.size(); i++) {
        tag->artist[i] = id3v2Tag->artist[i].c_str();
    }

    return tag;
}

void freeID3v2Tag(const ID3v2Tag *tag) {
    free(tag->artist);
    delete tag->_owner->tag;
    delete tag->_owner;
    delete tag;
}

// FFI_PLUGIN_EXPORT void printID3v2Tag(ID3v2Tag *id3V2Tag) {
//     if (id3V2Tag == nullptr) {
//         std::cout << "ptr is empty" << std::endl;
//         return;
//     }
//
//     std::cout << "File: " << id3V2Tag->fileName << "\n"
//               << "File Path: " << id3V2Tag->filePath << "\n"
//               << "Title: " << id3V2Tag->title << "\n"
//               << "Artist: " << id3V2Tag->artist << "\n"
//               << "Album Title: " << id3V2Tag->albumTitle << "\n"
//               << "Album Artist: " << id3V2Tag->albumArtist << "\n"
//               << "Year: " << id3V2Tag->year << "\n"
//               << "Track: " << id3V2Tag->track << "\n"
//               << "Album Track Count: " << id3V2Tag->albumTotalTrack << "\n"
//               << "Genre: " << id3V2Tag->genre << "\n"
//               << "Comment: " << id3V2Tag->comment << "\n"
//               << "Bit Rate: " << id3V2Tag->bitRate << "\n"
//               << "Sample Rate: " << id3V2Tag->sampleRate << "\n"
//               << "Channels: " << id3V2Tag->channels << "\n"
//               << "LengthInSeconds: " << id3V2Tag->lengthInSeconds << "\n"
//               << "LengthInMilliseconds: " << id3V2Tag->lengthInMilliSeconds << "\n"
//               << "Album Cover (size): " << id3V2Tag->albumCoverLength << "\n"
//               << "Lyrics (size): " << id3V2Tag->lyricsLength << std::endl;
// }
